clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/ChainResolver.java

```diff
@@ -131,4 +133,79 @@ public ChainResolver(final OperationsCodec<K, V> codec, Expiry<? super K, ? supe
       return new ResolvedChain.Impl<>(chain, key, result, 0, expirationTime);
     }
   }
+
+  @SuppressWarnings("unchecked")
+  public Chain compact(Chain chain, long now) {
+    Map<K, PutOperation<K, V>> compacted = new HashMap<>(1);
+    for (Element element : chain) {
+      ByteBuffer payload = element.getPayload();
+      Operation<K, V> operation = codec.decode(payload);
+      K key = operation.getKey();
+      final Result<V> previousResult = compacted.get(key);
+      final Result<V> result = operation.apply(previousResult);
+      if(result == null) {
+        compacted.remove(key);
+      } else if (expiry == Expirations.noExpiration()) {
+        if (result instanceof PutOperation<?, ?>) {
+          compacted.put(key, (PutOperation<K, V>) result); //unchecked
+        } else {
+          compacted.put(key, new PutOperation<>(key, result.getValue(), -Long.MAX_VALUE));
+        }
+      } else {
+        if(operation.isExpiryAvailable()) {
+          long expirationTime = operation.expirationTime();
+          if (now >= expirationTime) {
+            compacted.remove(key);
+          } else if (result instanceof PutOperation<?, ?>) {
+            compacted.put(key, (PutOperation<K, V>) result); //unchecked
+          } else {
+            compacted.put(key, new PutOperation<>(key, result.getValue(), -expirationTime));
+          }
+        } else {
+          try {
+            if(previousResult == null) {
+              Duration duration = expiry.getExpiryForCreation(key, result.getValue());
```

Comment:
I'm pretty sure this can be refactored into something more palatable. For instance, line 167 and line 182 blocks seems to be copy&pasted

Reply:
Eliminated that duplication.  More might be possible, let me know what you think.